fix bug where some pkg/layers diagnostics are only filled if corresponding (non-layer) diagnostic are filled

antnguyen13 opened the pull request:
## What changes does this PR introduce?
Bug fix

## What is the current behaviour? 
The calculation of total tendencies for theta/salt and passing of these to pkg/layers will be non-zeroes only if the diagnostics to TOT[T,S]TEND are filled.  If not, zeros are passed to pkg/layers, variable layers_tottend is filled with zeros inside subroutine layers_fill inside layers_save.F and diagnostics LTto[1SLT,2TH,3RHO] and LSto[1SLT,2TH,3RHO] are all zeros. Same for diagnostics La[T,S]z if implicit diffusion is used and not DrFI_[TH ,SLT], and surface terms La[T,S]s if [T,S]FLUX are not diagnosed.

## What is the new behaviour 
Change logic to also compute diagnostics in question when LAYERS_THERMODYNAMICS-code is active indedepend of diagnostics variables.

## Does this PR introduce a breaking change? 
None.

## Other information:


## Suggested addition to `tag-index`
- pkg/layers: fix bug that leads to zero output for some of the `LAYERS_THERMODYNAMCS` diagnostics (LaTz, LaSz, LTto, LSto,LaTs,LaSs)
 o new parameter layers_useThermo allows calling s/r layers_fill only when it is necessary

antnguyen13 committed:
separate diagnostics_is_on check of tot[T,S]tend from useLayers

antnguyen13 committed:
Merge remote-tracking branch 'origin' master into layers_diag_tottend

mjlosch commented on pkg/diagnostics/diagnostics_fill_state.F:
```diff
@@ -371,15 +370,15 @@ SUBROUTINE DIAGNOSTICS_FILL_STATE( selectVars, myIter, myThid )
            ENDDO
           ENDDO
          ENDDO
+        IF ( DIAGNOSTICS_IS_ON('TOTTTEND',myThid) ) THEN
          CALL DIAGNOSTICS_FILL(tmpMk,'TOTTTEND',0,Nr,0,1,1,myThid)
+        ENDIF
```
I am not sure, but if there are not additional computations associated with this diagnostic, then you don't need this if-statement any longer. After remove the if-statement, the entire block needs new indenting, right? Same for the cases below.

antnguyen13 replied:
I thought the if statement here is needed in the case if you do not want to fill TOT[T,S]TEND (so false here), such that you would not call to fill them?

mjlosch replied:
You are right, sorry, shouldn't do this in the middle of the night.

antnguyen13 committed:
add ifdef LAYERS_DIAG_TOTTEND and logical layers_diag_tottendTS

The values for tottend[T,S] are calculated when
(1) diagnostics of tottend[T,S] are requested outside of pkg/layers
AND/OR
(2) diagnostics of tottend[T,S] are requested INSIDE pkg/layers.
In case of (2) inside pkg/layers, prior to being diagnosed, the
tottend variables are being passed around as input arguments for
subroutines layers_calc.F and layers_fluxcalc.F, prior to the final
diagnostic call. Here, we introduce (a) option LAYERS_DIAG_TOTTEND
in LAYERS_OPTIONS.h (default to #define) to strip all array definitions,
subroutine arguments, and related calculations, thus saving memory
and computed time, and (b) logical "layers_diag_tottendTS" (default
to .TRUE.) to allow the filling of the layers diagnotics.

antnguyen13 commented:
@jm-c I made some changes, as you will see, by introducing two new things:
1. an option to allow bypassing defining of large arrays and calculations of these arrays (if undef) in the case we dont care to diagnose layers tottend
2. a logical to finally determine if we really write out layers diagnostics.

The reason for 1 is explained above: even if we don't want to diagnose, we're still passing around very large arrays and computing things that are all zeros.

The reason for 2, instead of just using diagnostic_is_on , is that the names of the diagnostics are not fixed, with a number "1" or "2" floating around depending on order of T/S (although it might have been hardcoded 1=T, 2=S, or the other way around... but I dont want to make a mess in diagnostics_fill_state.F to determine this string, which requiring several other pkg/layers variables.. so i make it easier by defining the associated/accompanied logical.

antnguyen13 committed:
remove recently added logical layers_diags_tottendTS and fix bug

Due to the way tottend variables are being used, defined, calculated,
and subsequently diagnosed within pkg/layers, the best approach is
to wrap it around the ifdef block as now done with previously
defined LAYERS_DIAG_TOTTEND compiling option without the need of
accompanied complexity of another logical.

Now, outside of pkg/layers, if uselayers is true, the tottend
is being computed and a call is done to pass it into pkg/layers.
However, inside pkg/layers, only within #ifdef LAYERS_DIAG_TOTTEND
block will any calculations and diagnostics relevant to tottend
will be done.

antnguyen13 commented:
@jm-c after looking more carefully at the way tottend variables are used in pkg/layers and how to call this from outside, i've simplified the problem by 

a. doing as you suggested in diagnostics_fill_state.F: simply hook in the additional check for useLayers.  Even though this variable is defined outside of pkg/layers, a subsequent call to routine layers_fill requires the additional wrap "#ifdef ALLOW_LAYERS" .
b. now, inside pkg/layers, everything related to tottend will only be defined, used, computed, and diagnosed if the new option LAYERS_DIAG_TOTTEND is defined.

antnguyen13 committed:
revert back any changes in pkg/layers

jm-c commented:
@antnguyen13 I tried to run "testreport -devel" using gfortran but got this error when running `cfc_example`:
> At line 6739 of file layers_fluxcalc.f
> Fortran runtime error: Index '0' of dimension 1 of array 'mybylo' below lower bound of 1
Could you check ?

antnguyen13 committed:
remove left-over unneeded CPP option

antnguyen13 commented:
@jm-c eeks, i'm appalled by my sloppiness.. now should be ok!

jm-c committed:
Merge branch 'master' into layers_diag_tottend

jm-c commented:
2 comments:

1. I really think it would be useful to open a new issue to list the problems related to `LAYERS_THERMODYNAMICS` code.
2. If we want to refine what you added in this PR: I took a quick look at this "total-tendency" layers bits but it's also output outside pkg/diagnostics, so it's not enough to check if the `pkg/layers` corresponding diagnostics is used. I might cook something simple and push it here.

jm-c committed:
New switch to record the use of Layers-Thermo part

- Register the use of Layers-Thermodynamics code by setting new switch "layers_useThermo",
  set at compile time (CPP option LAYERS_THERMODYNAMICS) since, currently, there is no
  run-time params for this ;
- Add a stop if Layers-Thermodynamics code is use without pkg/diagnostics (since all
  the calls are within "IF ( useDiagnostics )" ;
Note: put new switch in (new) dedicated header file "LAYERS_SWITCH.h" to facilitate
      access from outside pkg/layers (no need to include LAYERS_OPTIONS.h)

jm-c committed:
use a more specific condition for pkg/layers calls

using the (new) "layers_useThermo" switch (instead of "useLayers") allow to refine
the condition and only call pkg/layers S/R when Layers-Thermodynamics code is used

also some minor changes that minimize differences with original version

jm-c commented:
@antnguyen13 I pushed my little changes. If you don't like these we could revert the 2 commits.

jm-c committed:
Merge branch 'master' into layers_diag_tottend

mjlosch commented:
As far as I can see, the same issue appears for the diagnostics 'LaT/Sz1RHO' (i.e. the vertical diffusive fluxes), if we use implicit vertical diffusion (so basically all the time with a mixed layer model turned on). All other fluxes appear to be OK (horizontal diffusion, horizontal and vertical advection), because the fluxes are always computed to step the model forward whereas the implicit diffusive fluxes (and the total tendencies) are diagnosed after the model has been stepped forward. I suggest to add the implicit diffusive flux code here, too, and I even volunteer to do this.

jm-c commented:
Regarding names: I am thinking of renaming `LAYERS_SWITCH.h` to `LAYERS_P2SHARE.h`, similar to `DIAGNOSTICS_P2SHARE.h` (which has a very similar purpose, to enable 1 parameter to be seen/used outside the package code itself). We could rename `layers_useThermo` to something else ; just as I don't really know what the code under CPP option `LAYERS_THERMODYNAMICS` is doing exactly, I've just pick a name that reflect the CPP option.

mjlosch commented:
We can use the new flag `layers_useThermo` to reduce computational overhead in even more routines: 

- `gad_advection.F`
- `gad_calc_rhs.F`
- `gad_implicit_r.F` (this actually needs fixing as much as `impldiff.F`)
- `gad_som_advect.F`

As these are not big incisions, I'd like to make them also here in this PR. There may be remaining issues with `pkg/layers` (see issues #44, #60, #825), which we can address separately, as they are not directly related to the general problem here (layers-diagnostics not being filled).

jm-c committed:
Merge branch 'master' into layers_diag_tottend

jm-c committed:
rename LAYERS_SWITCH.h to LAYERS_P2SHARE.h

jm-c committed:
update description

jm-c commented:
@mjlosch I agree with your plan.

mjlosch committed:
use new layers_useThermo to fix similar issues in impldiff code

- in impldiff.F and gad_implicit_r.F
- avoid calling S/R LAYERS_FILL code if not layers_useThermo
- with this commit I can produce non-zero output for LaTz/LaSz,
  LTto/LSto (and as before all other "LAYERS_THERMODYNAMICS" related
  diagnostics variables) without tricks in data.diagnostics in
  cfc_example; whether this is all correct is a different story.

mjlosch commented:
With the latest commit [504e7df](https://github.com/MITgcm/MITgcm/pull/720/commits/504e7df9217ef53385e5f54776158fe972e4251b)  I can produce non-zero output for vertical diffusion terms `LaTz`, `LaSz` and total tendency terms  `LTto`, `LSto` (and as before all other "LAYERS_THERMODYNAMICS" related diagnostics variables) without tricks in data.diagnostics (i.e. without also saving `DrFI_TH /SLT` and `TOTT/STEND`) in `cfc_example` (the only place where we test this code). Whether this is all correct is a different story, but I think that now this PR is complete (provided I didn't introduce new error with my commit) and we should try to merge this. @antnguyen13 do you agree with this?

We would need to modify the description. I will add a tag-index entry suggestion.

mjlosch committed:
fix same bug for surface fluxes

mjlosch committed:
typo and another instance of replacing useLayers by layers_useThermo

jm-c committed:
minor: tweak indentation

jm-c reviewed (approved):
LGTM

jm-c committed:
Merge branch 'master' into layers_diag_tottend

jm-c committed:
document a fix for LAYERS_THERMODYNAMICS code
